File: RxTools-library/src/main/java/com/zbar/lib/FlashlightManager.java
package com.zbar.lib;

import android.os.IBinder;
import android.util.Log;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 *
 * 描述: 闪光灯管理
 */
final class FlashlightManager {

  private static final String TAG = FlashlightManager.class.getSimpleName();

  private static final Object iHardwareService;
  private static final Method setFlashEnabledMethod;
  static {
    iHardwareService = getHardwareService();
    setFlashEnabledMethod = getSetFlashEnabledMethod(iHardwareService);
    if (iHardwareService == null) {
      Log.v(TAG, "This device does supports control of a flashlight");
    } else {
      Log.v(TAG, "This device does not support control of a flashlight");
    }
  }

  private FlashlightManager() {
  }

  private static Object getHardwareService() {
    Class<?> serviceManagerClass = maybeForName("android.os.ServiceManager");
    if (serviceManagerClass == null) {
      return null;
    }

    Method getServiceMethod = maybeGetMethod(serviceManagerClass, "getService", String.class);
    if (getServiceMethod == null) {
      return null;
    }

    Object hardwareService = invoke(getServiceMethod, null, "hardware");
    if (hardwareService == null) {
      return null;
    }

    Class<?> iHardwareServiceStubClass = maybeForName("android.os.IHardwareService$Stub");
    if (iHardwareServiceStubClass == null) {
      return null;
    }

    Method asInterfaceMethod = maybeGetMethod(iHardwareServiceStubClass, "asInterface", IBinder.class);
    if (asInterfaceMethod == null) {
      return null;
    }

    return invoke(asInterfaceMethod, null, hardwareService);
  }

  private static Method getSetFlashEnabledMethod(Object iHardwareService) {
    if (iHardwareService == null) {
      return null;
    }
    Class<?> proxyClass = iHardwareService.getClass();
    return maybeGetMethod(proxyClass, "setFlashlightEnabled", boolean.class);
  }

  private static Class<?> maybeForName(String name) {
    try {
      return Class.forName(name);
    } catch (ClassNotFoundException cnfe) {
      // OK
      return null;
    } catch (RuntimeException re) {
      Log.w(TAG, "Unexpected error while finding class " + name, re);
      return null;
    }
  }

  private static Method maybeGetMethod(Class<?> clazz, String name, Class<?>... argClasses) {
    try {
      return clazz.getMethod(name, argClasses);
    } catch (NoSuchMethodException nsme) {
      // OK
      return null;
    } catch (RuntimeException re) {
      Log.w(TAG, "Unexpected error while finding method " + name, re);
      return null;
    }
  }

  private static Object invoke(Method method, Object instance, Object... args) {
    try {
      return method.invoke(instance, args);
    } catch (IllegalAccessException e) {
      Log.w(TAG, "Unexpected error while invoking " + method, e);
      return null;
    } catch (InvocationTargetException e) {
      Log.w(TAG, "Unexpected error while invoking " + method, e.getCause());
      return null;
    } catch (RuntimeException re) {
      Log.w(TAG, "Unexpected error while invoking " + method, re);
      return null;
    }
  }

  static void enableFlashlight() {
    setFlashlight(true);
  }

  static void disableFlashlight() {
    setFlashlight(false);
  }

  private static void setFlashlight(boolean active) {
    if (iHardwareService != null) {
      invoke(setFlashEnabledMethod, iHardwareService, active);
    }
  }

}

File: RxFeature/src/main/java/com/tamsiree/rxfeature/scaner/FlashlightManager.kt
package com.tamsiree.rxfeature.scaner

import android.os.IBinder
import com.tamsiree.rxkit.TLog.v
import com.tamsiree.rxkit.TLog.w
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

/**
 * @author Tamsiree
 */
internal object FlashlightManager {
    private val TAG = FlashlightManager::class.java.simpleName
    private var iHardwareService: Any? = null
    private var setFlashEnabledMethod: Method? = null

    @JvmStatic
    fun enableFlashlight() {
        setFlashlight(false)
    }

    @JvmStatic
    fun disableFlashlight() {
        setFlashlight(false)
    }

    private val hardwareService: Any?
        private get() {
            val serviceManagerClass = maybeForName("android.os.ServiceManager") ?: return null
            val getServiceMethod = maybeGetMethod(serviceManagerClass, "getService", String::class.java)
                    ?: return null
            val hardwareService = invoke(getServiceMethod, null, "hardware") ?: return null
            val iHardwareServiceStubClass = maybeForName("android.os.IHardwareService\$Stub")
                    ?: return null
            val asInterfaceMethod = maybeGetMethod(iHardwareServiceStubClass, "asInterface", IBinder::class.java)
                    ?: return null
            return invoke(asInterfaceMethod, null, hardwareService)
        }

    private fun getSetFlashEnabledMethod(iHardwareService: Any?): Method? {
        if (iHardwareService == null) {
            return null
        }
        val proxyClass: Class<*> = iHardwareService.javaClass
        return maybeGetMethod(proxyClass, "setFlashlightEnabled", Boolean::class.javaPrimitiveType!!)
    }

    private fun maybeForName(name: String): Class<*>? {
        return try {
            Class.forName(name)
        } catch (cnfe: ClassNotFoundException) {
            // OK
            null
        } catch (re: RuntimeException) {
            w(TAG, "Unexpected error while finding class $name", re)
            null
        }
    }

    private fun maybeGetMethod(clazz: Class<*>, name: String, vararg argClasses: Class<*>): Method? {
        return try {
            clazz.getMethod(name, *argClasses)
        } catch (nsme: NoSuchMethodException) {
            // OK
            null
        } catch (re: RuntimeException) {
            w(TAG, "Unexpected error while finding method $name", re)
            null
        }
    }

    private operator fun invoke(method: Method?, instance: Any?, vararg args: Any): Any? {
        return try {
            method!!.invoke(instance, *args)
        } catch (e: IllegalAccessException) {
            w(TAG, "Unexpected error while invoking $method", e)
            null
        } catch (e: InvocationTargetException) {
            w(TAG, "Unexpected error while invoking $method", e.cause)
            null
        } catch (re: RuntimeException) {
            w(TAG, "Unexpected error while invoking $method", re)
            null
        }
    }

    private fun setFlashlight(active: Boolean) {
        if (iHardwareService != null) {
            invoke(setFlashEnabledMethod, iHardwareService, active)
        }
    }

    init {
        iHardwareService = hardwareService
        setFlashEnabledMethod = getSetFlashEnabledMethod(iHardwareService)
        if (iHardwareService == null) {
            v(TAG, "This device does supports control of a flashlight")
        } else {
            v(TAG, "This device does not support control of a flashlight")
        }
    }
}